src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,334 @@
+        principals: this.permissions![permissionType],
+        users: principals.users,
+        groups: principals.groups,
+      });
+      if (result) {
+        this.permissions[permissionType] = result;
+      }
+    }
+
+    return this;
+  }
+
+  /**
+   * A function that transforms permissions format, change the format from permissionType->principals to principal->permissionTypes,
+   * which is used to clearyly dispaly user/group list and their granted permissions in the UI
+   *
+   * for example:
+   * the original permissions object is:   {
+   *     read: {
+   *         users:['user1']
+   *     },
+   *     write:{
+   *         groups:['group1']
+   *     }
+   * }
+   *
+   * the transformed permissions object will be: [
+   *     {type:'users', name:'user1', permissions:['read']},
+   *     {type:'groups', name:'group1', permissions:['write']},
+   * ]
+   *
+   * @returns the flat list of the permissions object
+   *
+   * @public
+   */
+  public toFlatList(): TransformedPermission[] {
+    const result: TransformedPermission[] = [];
+    if (!this.permissions) {
+      return result;
+    }
+
+    for (const permissionType in this.permissions) {
+      if (Object.prototype.hasOwnProperty.call(this.permissions, permissionType)) {
+        const { users = [], groups = [] } = this.permissions[permissionType] ?? {};
+        users.forEach((user) => {
+          const found = result.find((r) => r.type === PrincipalType.Users && r.name === user);
+          if (found) {
+            found.permissions.push(permissionType);
+          } else {
+            result.push({ type: PrincipalType.Users, name: user, permissions: [permissionType] });
+          }
+        });
+        groups.forEach((group) => {
+          const found = result.find((r) => r.type === PrincipalType.Groups && r.name === group);
+          if (found) {
+            found.permissions.push(permissionType);
+          } else {
+            result.push({ type: PrincipalType.Groups, name: group, permissions: [permissionType] });
+          }
+        });
+      }
+    }
+
+    return result;
+  }
+
+  /**
+   * A permissions object build function that resets the permissions object
+   *
+   * @public
+   */
+  public resetPermissions() {
+    // reset permissions
+    this.permissions = {};
+  }
+
+  /**
+   * A function that gets the premissions object
+   *
+   * @public
+   */
+  public getPermissions() {
+    return this.permissions;
+  }
+
+  /**
+   * A function that generates query DSL by the specific conditions, used for fetching saved objects from the saved objects index
+   *
+   * @param {Array} permissionTypes the permission types
+   * @param {Object} principals the users or groups
+   * @param {String | Array} savedObjectType saved object type, such as workspace, index-pattern etc.
+   * @returns the generated query DSL
+   *
+   * @public
+   * @static
+   */
+  public static generateGetPermittedSavedObjectsQueryDSL(
```

Comment:
where is this function being used?

Reply:
You can see the caller in our dev branch: [getQueryParams()](https://github.com/ruanyl/OpenSearch-Dashboards/blob/29ccc203b591141aaf4486e361ec31acf1b93be4/src/core/server/saved_objects/service/lib/search_dsl/query_params.ts#L290), it's used when generating a query DSL to find all or some specified saved objects which users/groups have permissions.